### ISSUE TITLE
Swift/iOS wrapper for TFLite libdeepspeech

### ISSUE THREAD

**Issue by reuben** [mozilla/DeepSpeech#3061]
WIP: Build works fine on TF r2.2 with some small modifications (TF branch [ios-build](https://github.com/mozilla/tensorflow/tree/ios-build) + DS branch [ios-build](https://github.com/mozilla/DeepSpeech/tree/ios-build)):

Build for simulator x86_64:

```bash
$ bazel build --verbose_failures --config=ios_x86_64 --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic -c opt //native_client:libdeepspeech.so --define=runtime=tflite --copt=-DTFLITE_WITH_RUY_GEMV
$ cp bazel-bin/native_client/libdeepspeech.so ../libdeepspeech.so.x86_64
```

Build for arm64:

```bash
$ bazel build --verbose_failures --config=ios_arm64 --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic -c opt //native_client:libdeepspeech.so --define=runtime=tflite --copt=-DTFLITE_WITH_RUY_GEMV
$ cp bazel-bin/native_client/libdeepspeech.so ../libdeepspeech.so.arm64
```

Create a fat binary with both archs:

```bash
$ cd ..
$ lipo -create libdeepspeech.so.arm64 libdeepspeech.so.x86_64 -output libdeepspeech.dylib
```

Then add it to Swift framework project in Xcode:

```bash
$ cp libdeepspeech.dylib DeepSpeech/native_client/swift/deepspeech_ios/
```

Then build the framework:

```bash
$ # Simulator:
$ xcodebuild -target deepspeech_ios -configuration Release -arch x86_64 defines_module=yes -sdk "iphonesimulator"
$ # Device:
$ xcodebuild -target deepspeech_ios -configuration Release -arch arm64 defines_module=yes -sdk "iphoneos"
```

Untested things:

- [ ] Does it actually run? Need to write a test app
- [ ] How to embed and load model in device/simulator
- [ ] How to test things in CI (can we embed a command line client binary in a test app and call it via simctl somehow to get stdout/stderr?)

**Comment by reuben** [comment]
It looks like building a fat binary for x86_64 and arm64 is not useful because you still need two separate frameworks for device and simulator, as they use different SDKs.

**Comment by reuben** [comment]
Some initial tests on device (iPhone Xs), averaged across 3 runs:

no scorer, cold cache: RTF 0.60x
no scorer, warm cache: RTF 0.48x
with scorer, cold cache: RTF 0.55x
with scorer, warm cache: RTF 0.24x

**Comment by reuben** [comment]
@lissyx could you give me a quick overview of what it would take to test changes to macOS workers in an isolated environment? Like, say, adding a macos-heavy-b worker type and spawning tasks with it to test.

**Comment by reuben** [comment]
My probably incomplete idea:
1. Make PR against https://github.com/mozilla/community-tc-config/blob/master/config/projects/deepspeech.yml adding new worker instance with -b type.
2. Wait for it to be landed and deployed.
3. Make a copy of one of the existing worker images, change the worker type, make modifications.
4. Spawn tasks against new worker type.

**Comment by lissyx** [comment]
> My probably incomplete idea:
> 
>     1. Make PR against https://github.com/mozilla/community-tc-config/blob/master/config/projects/deepspeech.yml adding new worker instance with -b type.
> 
>     2. Wait for it to be landed and deployed.
> 
>     3. Make a copy of one of the existing worker images, change the worker type, make other modifications, start VM.
> 
>     4. Spawn tasks against new worker type.

You'd have to use the script that is on the worker #1 `prov.sh`, and you'd have to update the base image prior to that because I have not done it:
 - `generic-worker` version
 - `taskclusterProxy`
 - `generic-worker.json` config update

We don't have a nicer way to spin new workers mostly because it's not something we needed to do often and because it'd require again much more tooling. Given the current status of our macOS workers ...

Doing that in parallel of running existing infra is likely to be complicated because of ... resources (CPU / RAM). I thought disk would be an issue but that should be fine.

**Comment by reuben** [comment]
> You'd have to use the script that is on the worker #1 `prov.sh`, and you'd have to update the base image prior to that because I have not done it:
> 
>     * `generic-worker` version
> 
>     * `taskclusterProxy`
> 
>     * `generic-worker.json` config update

I assume I can find the appropriate versions and config changes by inspecting the currently running VMs, right? In that case, I just need the IP of worker 1 to get started.

> We don't have a nicer way to spin new workers mostly because it's not something we needed to do often and because it'd require again much more tooling. Given the current status of our macOS workers ...

Yeah. I thought about making a VM copy of a worker to side-step these provisioning issues but I guess that's also prone to causing problems.

> Doing that in parallel of running existing infra is likely to be complicated because of ... resources (CPU / RAM). I thought disk would be an issue but that should be fine.

I would probably run that worker on my personal machine while I test it, since it's not meant for general availability.

**Comment by lissyx** [comment]
> I assume I can find the appropriate versions and config changes by inspecting the currently running VMs, right? In that case, I just need the IP of worker 1 to get started.

Indeed, you can fetch the json config. IPs on matrix.


> I would probably run that worker on my personal machine while I test it, since it's not meant for general availability.

Be aware the existing workers if you copy them to your system are meant for VMWare Fusion Pro

**Comment by reuben** [comment]
> Be aware the existing workers if you copy them to your system are meant for VMWare Fusion Pro

Ah, yes, that was also one of the complications. OK, thanks.

**Comment by reuben** [comment]
(Hopefully) finished wrapping the C API, now moving on to CI work. Wrapper is here if anyone wants to take a quick look and provide any suggestions: https://github.com/mozilla/DeepSpeech/blob/ios-build/native_client/swift/deepspeech_ios/DeepSpeech.swift

**Comment by reuben** [comment]
In particular I'd be very interested in any feedback from Swift developers on how the error handling looks and how the buffer handling around `Model.speechToText` and `Stream.feedAudioContent` looks.

**Comment by reuben** [comment]
PR for adding macos-heavy-b and macos-light-b worker type and instances: https://github.com/mozilla/community-tc-config/pull/308

**Comment by reuben** [comment]
Looks like adding Xcode to the worker brings the free space to under 10GB which stops the taskcluster client from picking up any jobs... Resizing the partition does not seem to work. My next step will be to create a worker from scratch with a larger disk image.

**Comment by reuben** [comment]
@dabinat tagging you because you mentioned interest in these bindings in the CoreML issue, in case you have anything to mention regarding the design of the bindings here.

**Comment by erksch** [comment]
@reuben super awesome that you are working on this! This is actually perfect timing as we are looking for an offline speech recognition for iOS right now. I know it's not finished yet, but could you provide a small guide on how I could try it out, is the .so library already available somewhere? Maybe then I could also help with writing an example app.

**Comment by lissyx** [comment]
> @reuben super awesome that you are working on this! This is actually perfect timing as we are looking for an offline speech recognition for iOS right now. I know it's not finished yet, but could you provide a small guide on how I could try it out, is the .so library already available somewhere? Maybe then I could also help with writing an example app, if you wish.

on taskcluster, if you browse to the iOS artifacts sections you should get it

**Comment by lissyx** [comment]
e.g., https://community-tc.services.mozilla.com/tasks/BO-_Adi2Th2DOswsw9m1Kw#artifacts

**Comment by erksch** [comment]
Thank you! I'll try it out!

**Comment by erksch** [comment]
I tried it out with the current 0.7.4 models from the release page and one of the audio files from there.

```
TensorFlow: v2.2.0-17-g0854bb5188
DeepSpeech: v0.9.0-alpha.2-34-gdd20d35c
2020-07-17 09:48:31.981587-0700 deepspeech_ios_test[9411:91040] Initialized TensorFlow Lite runtime.
/private/var/containers/Bundle/Application/F5F8492E-D9B8-4BC4-AF46-29CEB23FC3A6/deepspeech_ios_test.app/4507-16021-0012.wav
read 8192 samples
(lldb)
```

And then comes this error 
```
Thread 5: EXC_BAD_ACCESS (code=1, address=0x177000075)
```
in this line of `DeepSpeech.swift`
```swift
178    public func feedAudioContent(buffer: UnsafeBufferPointer<Int16>) {
180        precondition(streamCtx != nil, "calling method on invalidated Stream")
181        
182        DS_FeedAudioContent(streamCtx, buffer.baseAddress, UInt32(buffer.count)) <<<<< Thread 5: EXC_BAD_ACCESS (code=1, address=0x177000075)
183   }
```

Just leaving this here, but I don't want to bother too much before this is even called finished :D

**Comment by reuben** [comment]
The models should be compatible. I don't know what's going on there, can't reproduce it locally...

**Comment by reuben** [comment]
Does the log have any more details for this error?

**Comment by reuben** [comment]
Also, maybe double check the signing options in Xcode? At some point when writing the bindings I ran into some runtime exceptions due to incorrect signing options.

**Comment by erksch** [comment]
You're right. Since it crashes when communication with library happens, it's probably just included incorrectly.

So do go through what I tried:
- Cloning the repo and checking out the ios-build branch
- In XCode, set signing for the targets `deepspeech_ios` and `deepspeech_ios_test` to my team and adjust the bundle identifier to one of mine
- Trying to run deepspeech_ios_test on my device
-> Build failed with
```
clang: error: no such file or directory: '[...]/DeepSpeech/native_client/swift/libdeepspeech.so'
Command Ld failed with a nonzero exit code
```
- So downloading the ARM library from the link that you provided and moving it to the given destination
- Trying to run again 
-> Build succeeded but runtime error
```
dyld: Library not loaded: @rpath/deepspeech_ios.framework/deepspeech_ios
  Referenced from: /private/var/containers/Bundle/Application/E9B900F3-5F4C-466D-BB03-E97F5588A768/deepspeech_ios_test.app/deepspeech_ios_test
  Reason: image not found
(lldb) bt 
* thread #1, stop reason = signal SIGABRT
```

The error above is after trying some things in the "Frameworks, Libraries and Embedded Content Section".
After doing a fresh start and just adding the library like explained above I get the following configs:
- `deepspeech_ios_test` target
<img width="741" alt="Bildschirmfoto 2020-07-19 um 21 10 15" src="https://user-images.githubusercontent.com/19290349/87882983-500e0f00-ca04-11ea-9166-6f5d815c2c49.png">

- `deepspeech_ios` target
<img width="743" alt="Bildschirmfoto 2020-07-19 um 21 11 03" src="https://user-images.githubusercontent.com/19290349/87882998-6d42dd80-ca04-11ea-8823-f9664020e705.png">

What do you have there? Maybe some things have to be switched to embed & sign?

**Comment by erksch** [comment]
After setting `deepspeech_ios.framework` to `Embed & Sign` in the `deepspeech_ios_test` target (which is just a random tryout), the code at least passes until the `DS_FeedAudioContent`, and the error occurs that I mentioned in the first post. 

Here is the full log I got for that error.
```
* thread #2, queue = 'com.apple.avfoundation.avasset.completionsQueue', stop reason = EXC_BAD_ACCESS (code=2, address=0x130800076)
    frame #0: 0x0000000103891494 libdeepspeech.so`___lldb_unnamed_symbol5$$libdeepspeech.so + 360
  * frame #1: 0x000000010385df68 deepspeech_ios`DeepSpeechStream.feedAudioContent(buffer=(_position = 0x000000016502d200, count = 8192), self=(streamCtx = 0x0000000162f32f40)) at DeepSpeech.swift:181:9
    frame #2: 0x00000001028b20b0 deepspeech_ios_test`closure #1 in render(samples=Swift.UnsafeRawBufferPointer @ 0x000000016d5e0100, stream=(streamCtx = 0x0000000162f32f40)) at AppDelegate.swift:126:20
    frame #3: 0x00000001028b2138 deepspeech_ios_test`thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #4: 0x00000001028b2198 deepspeech_ios_test`partial apply for thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #5: 0x00000001b810c348 libswiftFoundation.dylib`Foundation.Data.withUnsafeBytes<A>((Swift.UnsafeRawBufferPointer) throws -> A) throws -> A + 504
    frame #6: 0x00000001028b097c deepspeech_ios_test`render(audioContext=0x000000016760bd10, stream=(streamCtx = 0x0000000162f32f40)) at AppDelegate.swift:124:22
    frame #7: 0x00000001028b30c8 deepspeech_ios_test`closure #1 in test(audioContext=0x000000016760bd10, stream=(streamCtx = 0x0000000162f32f40), audioPath="/private/var/containers/Bundle/Application/D6D001A2-07F7-4BD3-80E9-9DBECCA975E8/deepspeech_ios_test.app/4507-16021-0012.wav", start=2020-07-19 21:19:15 CEST, completion=0x00000001028b83c8 deepspeech_ios_test`partial apply forwarder for closure #1 () -> () in closure #1 () -> () in deepspeech_ios_test.AppDelegate.application(_: __C.UIApplication, didFinishLaunchingWithOptions: Swift.Optional<Swift.Dictionary<__C.UIApplicationLaunchOptionsKey, Any>>) -> Swift.Bool at <compiler-generated>) at AppDelegate.swift:174:9
    frame #8: 0x00000001028acfc0 deepspeech_ios_test`closure #1 in static AudioContext.load(asset=0x000000016365fd40, assetTrack=0x00000001637212f0, audioURL=Foundation.URL @ 0x0000000163674090, completionHandler=0x00000001028b375c deepspeech_ios_test`partial apply forwarder for closure #1 (Swift.Optional<deepspeech_ios_test.AudioContext>) -> () in deepspeech_ios_test.test(model: deepspeech_ios.DeepSpeechModel, audioPath: Swift.String, completion: () -> ()) -> () at <compiler-generated>) at AppDelegate.swift:59:17
    frame #9: 0x00000001028ad9b0 deepspeech_ios_test`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0
    frame #10: 0x0000000102c0fefc libclang_rt.asan_ios_dynamic.dylib`__wrap_dispatch_async_block_invoke + 196
    frame #11: 0x0000000104ec605c libdispatch.dylib`_dispatch_call_block_and_release + 32
    frame #12: 0x0000000104ec74d8 libdispatch.dylib`_dispatch_client_callout + 20
    frame #13: 0x0000000104ecec20 libdispatch.dylib`_dispatch_lane_serial_drain + 720
    frame #14: 0x0000000104ecf834 libdispatch.dylib`_dispatch_lane_invoke + 440
    frame #15: 0x0000000104edb270 libdispatch.dylib`_dispatch_workloop_worker_thread + 1344
    frame #16: 0x00000001816a7718 libsystem_pthread.dylib`_pthread_wqthread + 276
```

**Comment by dabinat** [comment]
I tried a simple test app where I loaded a pre-converted file of a few seconds into memory and called DeepSpeechModel.speechToText. There are no crashes or anything, but the resulting text string is empty.

It seemed from the header like all I had to do was initialize a DeepSpeechModel with the .tflite file and then call speechToText with the buffer. Did I miss a step? Do I need to setup a streaming context even if I’m not streaming?

**Comment by reuben** [comment]
> It seemed from the header like all I had to do was initialize a DeepSpeechModel with the .tflite file and then call speechToText with the buffer. Did I miss a step? Do I need to setup a streaming context even if I’m not streaming?

That's correct, you don't need to setup a streaming context.

**Comment by reuben** [comment]
> What do you have there? Maybe some things have to be switched to embed & sign?

I have `libdeepspeech.so` in "Link Binary With Libraries" for project `deepspeech_ios`. I don't see a "Frameworks and Libraries" section like in your screenshot.

In `deepspeech_ios_test`, I have both `libdeepspeech.so` and `deepspeech_ios.framework` in "Link Binary With Libraries", and I also have `libdeepspeech.so` in "Embed Frameworks", with "Code Sign On Copy" enabled.

**Comment by reuben** [comment]
Oh, OK, I was looking at the "Build Phases" tab, not "General". What I have in "General" matches your screenshots. It's weird that it crashes on DS_FeedAudioContent and not DS_CreateModel, maybe it's not a linking issue, unless your code has a bug and it's calling DS_FeedAudioContent on an invalid model/stream, which would certainly cause a segfault.

**Comment by erksch** [comment]
You're right, DS_CreateModel seems to work fine and I can retrieve sample rate and beam width. So the issue for me has to be in sampling and memory allocation. It's weird that that part is not working for me somehow.

**Comment by erksch** [comment]
Okay weirdly enough, when I run the app multiple times, every now and then it continues feeding audio content successfully
```
read 8192 samples
read 8192 samples
read 8192 samples
read 8192 samples
read 8192 samples
read 2800 samples
```
But it always fails when calling `DS_FinishStream(streamCtx)` with a `EXC_BAD_ACCESS`.

But every other time it fails feeding the first batch of samples.

**Comment by reuben** [comment]
Out of memory maybe? What device are you running this on?

**Comment by reuben** [comment]
The iOS build has been added to CI in #3150. So we should now have automatically built binaries that can be used with the project. Next step is building and publishing the wrapper itself, but before doing that I want to get more feedback from users, such as these memory issues... (Thanks for testing @erksch !)

**Comment by erksch** [comment]
@reuben Hey just tried again from the current master, still the same. I am testing on brand new iPad Air with I think at least 4 GB RAM.

**Comment by erksch** [comment]
@reuben Btw, today I implemented microphone streaming on iOS with resampling to 16000Hz. I tested it with a different speech recognition but I think the code would transfer 100% to DeepSpeech. Do you think we could add this to the demo code?

**Comment by reuben** [comment]
@erksch I'm running on an iPhone Xs which is also 4GB of RAM, so that's weird.

Microphone streaming would be amazing to have! Can you open a PR?

**Comment by erksch** [comment]
Yep hopefully tomorrow :)

**Comment by reuben** [comment]
@erksch have you had a chance to work on the microphone streaming code? I would love to include that in 0.8 if possible :)

**Comment by erksch** [comment]
@reuben sorry for the absence, i was a little bit occupied. I'll make a PR today!

**Comment by erksch** [comment]
@reuben I noticed "/deepspeech_ios/libdeepspeech.dylib" in the .gitignore. Maybe that is what is missing for me? Where do I get this file?

**Comment by reuben** [comment]
@erksch that is a leftover from older code, I'll remove it. You should only need `libdeepspeech.so`.

**Comment by erksch** [comment]
[Here is the PR](https://github.com/mozilla/DeepSpeech/pull/3189)
While file transcription does still not work, the microphone streaming actually works for me :D

**Comment by zaptrem** [comment]
Do you plan on publishing a Cocoapod for the framework portion of this? I'm going to build a React Native module for this but auto-linking requires a pod. Afterward, I can add instructions for users to drag the model file into the correct directory.

**Comment by reuben** [comment]
It's not just the models, you also need to add the shared library, `libdeepspeech.so`, as a direct dependency of the app. Do you think it still makes sense to publish the framework on Cocoapods given this requirement? I've been hesitant to do that and then get tons of support requests from people with weird linker errors.

We could try to `dlopen` the library and give a more descriptive message but I'm not sure how that fits with App Store guidelines.

**Comment by zaptrem** [comment]
I'll look into that. Loading dynamic code from the internet isn't allowed on the AppStore.

**Comment by reuben** [comment]
I'm not talking about loading dynamic code from the internet, I'm talking about having to add a shared library as a dependency or your app in Xcode. It'll then get codesigned and bundled with the application. This is because iOS does not allow linking against a framework that itself links against another shared library. Maybe I should look more seriously into static linking, which would make this way simpler.

**Comment by zaptrem** [comment]
If that won't require some massive overhaul it might be a good idea. Static frameworks are sometimes better for app load times than dynamic ones. Let me know how this goes, as switching from Apple's wildly unstable built-in transcription API to DeepSpeech for extremely long on-device transcription jobs is on our roadmap (after trying it on Android starting in a few days).

**Comment by zaptrem** [comment]
> I'm not talking about loading dynamic code from the internet, I'm talking about having to add a shared library as a dependency or your app in Xcode. It'll then get codesigned and bundled with the application. This is because iOS does not allow linking against a framework that itself links against another shared library. Maybe I should look more seriously into static linking, which would make this way simpler.
> 
> 
> > Am 30.07.2020 um 22:28 schrieb zaptrem <notifications@github.com>:
> > 
> > ﻿
> > I'll look into that. Loading dynamic code from the internet isn't allowed on the AppStore.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> 

Any change on this? I'm starting the Android half of this library tonight.

**Comment by alexmay23** [comment]
@reuben hey. Where can I find libdeepspeech.so for iOS, or how can I build it by myself? I looked at links you have provided long time ago but it sends me 404. Thanks in advance.

**Comment by lissyx** [comment]
It's provided in the latest 0.8 or alpha 0.9 on github releases

**Comment by alexmay23** [comment]
@lissyx iOS build only in 0.9 alpha.

**Comment by lissyx** [comment]
> @lissyx iOS build only in 0.9 alpha.

It's always nice when you trust people that are working on the project.

**Comment by lissyx** [comment]
https://community-tc.services.mozilla.com/tasks/index/project.deepspeech.deepspeech.native_client.v0.8.1/ios_arm64-tflite
https://community-tc.services.mozilla.com/tasks/index/project.deepspeech.deepspeech.native_client.v0.8.1/ios_x86_64-tflite

**Comment by alexmay23** [comment]
> https://community-tc.services.mozilla.com/tasks/index/project.deepspeech.deepspeech.native_client.v0.8.1/ios_arm64-tflite
> https://community-tc.services.mozilla.com/tasks/index/project.deepspeech.deepspeech.native_client.v0.8.1/ios_x86_64-tflite

awesome. thank you.

**Comment by zaptrem** [comment]
@reuben I'm going over the test project you made for iOS. Why did you decide to use AVCaptureSession and all of its intricacies instead of AVAudioEngine? In addition to working in the background, it seem like fewer raw, unsafe pointers and mallocs would be needed to accomplish the same goal. I threw together something similar for a half-finished RN Module that uses Apple's SFSpeech AudioBuffer Recognizer a while ago: https://github.com/zaptrem/react-native-transcript/blob/master/ios/TranscriptLive.swift (you can ignore anything that mentions caching, Apple doesn't allow the Speech Recognizer to be started while in background so if it stops due to an error I cache the buffer until the user resumes the app). 

The summary is I create an AVAudioEngine instance, create a Recognizer, installTap on the engine's inputNode, and append the buffer to the Recognizer in the installTap's callback.

**Comment by erksch** [comment]
@zaptrem The microphone transcription was written by me. It works, but if you have a better idea feel free to make a new PR. Maybe I missed AVAudioEngine in my research on processing audio data.

> Also, is there a way to receive live transcription results like the android_mic_streaming example?

What do you mean by that? The iOS test app supports live microphone transcription.

**Comment by zaptrem** [comment]
@erksch I'm not getting any live microphone transcription updates as I record with the demo. Only after I tap "Stop Microphone Recognition" does a transcription print in the console.

I'm working on integrating the Android native client into a React Native module I'm building right now. After I finish that I'll start working with this again.

**Comment by zaptrem** [comment]
> @zaptrem I have not tried out closing the app when running the model. Is that a requirement for you?
> 
> > Instead, the library would stick with SFSpeechRecognizer for live recognitionInstead, the library would stick with SFSpeechRecognizer for live recognition
> 
> Actually, I would really like to see a library that is completely powered by DeepSpeech. Many people turn to DeepSpeech because of privacy concerns and offline inference. If this is not important to me I'd use `react-native-voice`.
@erksch 
I'm using it for long-running simultaneous recording+transcription jobs that would be price (and in some cases privacy) prohibitive in the cloud. Users likely can't/won't keep their phones unlocked for two hour recording sessions, so it is a requirement. SFSpeechRecognizer has an offline-only mode, but its far-field accuracy sucks and it is extremely unreliable (i.e., the recognizer straight up fails in the middle of jobs constantly). Once you guys start mixing in reverb, noise, and different loudness levels for 1.0, DeepSpeech will hopefully perform better (especially because from my limited testing in Android you're stable for long jobs and your token timestamps actually seem to be accurate, unlike Apple's). 

The React Native library, `react-native-transcription` will be open source (I don't want to publish it until the Android half is a little more fleshed-out). However, I'm using it as a backend for [a new class collaboration app, ReLearn](https://relearn.fyi). The Android version (going live for testing in a few days) is powered 100% by DeepSpeech, while the iOS version will stick with SFSpeechRecognizer for live recording until DeepSpeech is noise/reverb-hardened and we can verify background stability.

**Comment by zaptrem** [comment]
@erksch I decided to post the repo a little earlier than expected [here](https://github.com/zaptrem/react-native-transcription). Right now it only has Android live transcription support. I pushed it because I ran into a memory leak that seems to be linked to the Java native client implementation. I made an issue [here](https://github.com/mozilla/STT/issues/3271).

**Comment by zaptrem** [comment]
@reuben Finished the Android portion of the module [here](https://github.com/zaptrem/react-native-transcription).  I totally understand if you haven't thought about this considering all of the craziness going on at Mozilla right now, but has there been any change on the Cocoapods front? I want to make sure I integrate the iOS DeepSpeech library the way it's supposed to be.

**Comment by reuben** [comment]
I don't know anything about Cocoapods so I'm afraid I won't be helpful for the near term. Contributions always welcome.

**Comment by reuben** [comment]
Also it looks like both bazel and tensorflow require changes before we can have a fully static build: https://github.com/tensorflow/tensorflow/issues/28388 https://github.com/bazelbuild/bazel/issues/1920

**Comment by zaptrem** [comment]
@reuben The deepspeech_ios.framework files your build system is distributing don't have Build for Distribution enabled. As a result, they're only compatible with the exact Swift compiler you used to build them. See [here](https://stackoverflow.com/questions/58654714/module-compiled-with-swift-5-1-cannot-be-imported-by-the-swift-5-1-2-compiler). It's a 30 second fix, but I have no clue how your massive build system works so I'm unqualified to open a PR. Also, it would be helpful for testing if your frameworks and libdeepspeech.so files were compiled as universal binaries, so I didn't have to switch them out every time I wanted to run on simulator instead of a real device.

**Comment by reuben** [comment]
I couldn't figure out how to build a universal framework because they have to be built against different SDKs, not just architectures. I'll take a look at "Build for Distribution".

**Comment by zaptrem** [comment]
@reuben Is the iOS 13.5 minimum deployment target intentional? 
![Screen Shot 2020-08-26 at 6 52 34 PM](https://user-images.githubusercontent.com/1612230/91364861-ac1b3e80-e7cd-11ea-87a3-7bb360b35649.png)

**Comment by zaptrem** [comment]
Turns out I _was_ close! The repo now has a working wav file transcription example. However, it appears the CoreML delegate isn't being used, as CPU usage was at 400% during the Wav transcription. I can't be certain as afaik there's no way to profile Neural Engine usage. GPU usage was at 0%.

**Comment by zaptrem** [comment]
@reuben Alright, as of the latest commit we have the first working cross-platform RN module for DeepSpeech. I haven't built iOS live microphone transcription yet as I have another pressing matter to attend to for now, but .wav file transcription is cross platform and live transcription works on Android. I made a few changes to the framework (setting a bunch of important fields that should have been public-gettable but weren't) that would be great if they were added to 1.0. 

Let me know what you think!

I built it to be as simple as possible, which means there are no model options available to users for now. I want to make sure the most people get the best experience right off the bat. Do you have any suggestions for tuning settings (e.g., alpha, beta, scorer) to reach that goal?

**Comment by zaptrem** [comment]
It seems like it's impossible to submit an app that includes the DeepSpeech RN module to App Store Connect and Test Flight. Depending on how you compile the .framework it either fails to Archive with missing symbols (despite running fine on a real device in Release mode), is unable to upload at all for a million different reasons, or uploads and is rejected for "ITMS-90426: Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it."

One solution on the internet suggested disabling bitcode, but that just caused a million upload errors related to misaligned segmentations of some sort. I've been working at this for days and I'm completely stumped! @reuben If you'd be able to bundle the framework in a cocoapod even without libdeepspeech.so that would be extremely helpful. I think my understanding of how linkers and compilers work isn't at the point yet where I can solve this problem without tripping over the solution after a significant amount of random trial (if ever).

**Comment by erksch** [comment]
@zaptrem Unfortunately, I also know very little about what it takes to bring this library to CocoaPods, but it is also in my best interest that submitting to the store works (but maybe not as urgent as for you). Maybe we can combine our efforts and get it to work, although @reuben seems to have greatest expertise in building the library for iOS. 

You can write me on Telegram under the same username maybe we can setup a little Sprint for that.

**Comment by reuben** [comment]
Some searching turned this up, which apparently needs to be used when building deepspeech_ios.framework: https://stackoverflow.com/a/36546996/346048

**Comment by reuben** [comment]
To be clear, you don't actually need to rebuild libdeepspeech.so yourself, you can use the copy from our CI infra. Building just the wrapper framework library should be easily doable from source, just open it with Xcode or use xcodebuild like we do in CI (see `do_deepspeech_ios_framework_build` in `taskcluster/tc-build-utils.sh`).

**Comment by zaptrem** [comment]
> To be clear, you don't actually need to rebuild libdeepspeech.so yourself, you can use the copy from our CI infra. Building just the wrapper framework library should be easily doable from source, just open it with Xcode or use xcodebuild like we do in CI (see `do_deepspeech_ios_framework_build` in `taskcluster/tc-build-utils.sh`).

Yes, I'm already doing that because the Metadata and Token arrays were (erroneously?) marked as `private(set)` instead of `public private(set)`.



> @zaptrem Unfortunately, I also know very little about what it takes to bring this library to CocoaPods, but it is also in my best interest that submitting to the store works (but maybe not as urgent as for you). Maybe we can combine our efforts and get it to work, although @reuben seems to have greatest expertise in building the library for iOS.
> 
> You can write me on Telegram under the same username maybe we can setup a little Sprint for that.

Just did.

**Comment by xiaoqunSun** [comment]
> > To be clear, you don't actually need to rebuild libdeepspeech.so yourself, you can use the copy from our CI infra. Building just the wrapper framework library should be easily doable from source, just open it with Xcode or use xcodebuild like we do in CI (see `do_deepspeech_ios_framework_build` in `taskcluster/tc-build-utils.sh`).
> 
> Yes, I'm already doing that because the Metadata and Token arrays were (erroneously?) marked as `private(set)` instead of `public private(set)`.
> 
> > @zaptrem Unfortunately, I also know very little about what it takes to bring this library to CocoaPods, but it is also in my best interest that submitting to the store works (but maybe not as urgent as for you). Maybe we can combine our efforts and get it to work, although @reuben seems to have greatest expertise in building the library for iOS.
> > You can write me on Telegram under the same username maybe we can setup a little Sprint for that.
> 
> Just did.



> It seems like it's impossible to submit an app that includes the DeepSpeech RN module to App Store Connect and Test Flight. Depending on how you compile the .framework it either fails to Archive with missing symbols (despite running fine on a real device in Release mode), is unable to upload at all for a million different reasons, or uploads and is rejected for "ITMS-90426: Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it."
> 
> One solution on the internet suggested disabling bitcode, but that just caused a million upload errors related to misaligned segmentations of some sort. I've been working at this for days and I'm completely stumped! @reuben If you'd be able to bundle the framework in a cocoapod even without libdeepspeech.so that would be extremely helpful. I think my understanding of how linkers and compilers work isn't at the point yet where I can solve this problem without tripping over the solution after a significant amount of random trial (if ever).



> It seems like it's impossible to submit an app that includes the DeepSpeech RN module to App Store Connect and Test Flight. Depending on how you compile the .framework it either fails to Archive with missing symbols (despite running fine on a real device in Release mode), is unable to upload at all for a million different reasons, or uploads and is rejected for "ITMS-90426: Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it."
> 
> One solution on the internet suggested disabling bitcode, but that just caused a million upload errors related to misaligned segmentations of some sort. I've been working at this for days and I'm completely stumped! @reuben If you'd be able to bundle the framework in a cocoapod even without libdeepspeech.so that would be extremely helpful. I think my understanding of how linkers and compilers work isn't at the point yet where I can solve this problem without tripping over the solution after a significant amount of random trial (if ever).



Have you found a solution?

My IOS app also encountered this problem when it was submitted to the store

**Comment by xiaoqunSun** [comment]
Direct use of the dynamic library "so" will be rejected when submitted to the apple store



I tried to build the static library ". a",     but it didn't make any progress. Can you help me

**Comment by zaptrem** [comment]
@xiaoqunSun Are you sure this is the issue?

**Comment by zaptrem** [comment]
@reuben Attempting to upload an archive of the DeepSpeech Test app in the 0.8.2 repo to App Store Connect results in an email saying:

```
ITMS-90426: Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it.
```

I think @xiaoqunSun might be right or close to it.

**Comment by xiaoqunSun** [comment]
> @reuben尝试在0.8.2存储库中将DeepSpeech Test应用程序的存档上传到App Store Connect时，结果为电子邮件：
> 
> ```
> ITMS-90426: Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it.
> ```
> 
> 我认为@xiaoqunSun可能正确或接近它。

Yes,   same error

**Comment by zaptrem** [comment]
Maybe it would be possible package libdeepspeech.so and its’ header files as a SEPERATE cocoapod that a hypothetical deepspeech_ios would depend on.

**Comment by xiaoqunSun** [comment]
I found a way to build static framework


copy      tf_cc_shared_object(
                    name = "libdeepspeech.so",
                    ...................
               )
rename  cc_library(
                    name = "static",
                    ...................
               )
add 
ios_static_framework(
    name = "deepspeech_ios",
    deps = [":static"],
    families = ["iphone","ipad"],
    minimum_os_version = "9.0",
)

**Comment by zaptrem** [comment]
@xiaoqunSun  does this actually work? AFAIK TensorFlow and Basel don't work with static libraries.

Until this issue is fixed this library is essentially unstable on iOS.

**Comment by zaptrem** [comment]
@reuben Is there any solution in sight for the Missing SwiftSupport Folder error when uploading a binary with this framework included to AppStore Connect? There’s a major feature depending on this working that is ready on Android but has been completely blocked on iOS for nearly a month now.

**Comment by zaptrem** [comment]
@lissyx @reuben @xiaoqunSun @erksch looks like the libdeepspeech.so file must be bundled with the framework somehow. You aren’t allowed to put dynamic libraries directly in the main project, they must be wrapped in a .framework. I spent today trying to do that but have no clue what I’m doing. I hope these can help.

https://developer.apple.com/forums/thread/125796
https://stackoverflow.com/questions/57755276/create-ios-framework-with-dylib
https://stackoverflow.com/questions/53818772/embedding-a-dylib-inside-a-framework-for-ios
Thx Adidas?: https://www.runtastic.com/blog/en/frameworks-ios/

**Comment by reuben** [comment]
> Yes, I'm already doing that because the Metadata and Token arrays were (erroneously?) marked as `private(set)` instead of `public private(set)`.

PRs are welcome.

**Comment by reuben** [comment]
I'm seeing conflicting answers in the discussion here: is the problem that dynamic libraries aren't allowed at all? Is it that the dynamic library must be inside a framework? The latter is probably easier to fix than the former. You should be able to set the deepspeech_ios project to also embed libdeepspeech.so. If that works, please send a PR with the relevant changes. For static builds it's a much more complicated situation as I mentioned above: https://github.com/mozilla/DeepSpeech/issues/3061#issuecomment-680839647

**Comment by zaptrem** [comment]
The sources I sent seem to indicate dynamic libraries are allowed
(otherwise, how would any TensorFlow app work on iOS?) but they must be
wrapped in a framework. I tried setting the project to also embed
libdeepspeech, but then it acted like the .so wasn’t included at all. I
tried embedding, copying to the bundle as a resource, and copying as a
“framework” and none seemed to work. One of the sources suggested using a
really low level tool to convert the file to a specific binary format to
insert it into a framework manually, but that seems like the sort of thing
that needs to be built into the TaskCluster.

**Comment by zaptrem** [comment]
@reuben Is it possible to configure your build process to embed libdeepspeech.so into a .framework to make Apple happy?

**Comment by reuben** [comment]
@zaptrem the CI build just calls Xcode with `xcodebuild`, so if you can make it work on Xcode.app, it will also work on the CI build.

**Comment by zaptrem** [comment]
@reuben I spent a few days weeks ago trying to figure that part out, but couldn't figure out how to make it recognize the correct header files/binary files/etc.

**Comment by zaptrem** [comment]
Did the fallout of the Mozilla restructuring include deprioritizing DeepSpeech (esp new platform support like iOS)? I'm trying to figure out whether iOS support is still happening.

**Comment by zaptrem** [comment]
@reuben  Can I pay to open a bug bounty for this? I’d love to ship the feature built on DeepSpeech that has been shelved for months now.

**Comment by CatalinVoss** [comment]
I made it work yesterday as a static framework with cocoa pods, though integrated into another code base for a fairly specific application. TestFlight went through. I think I can turn it into a PR for you this week. Send some good karma.

**Comment by zaptrem** [comment]
> I made it work yesterday as a static framework with cocoa pods, though integrated into another code base for a fairly specific application. TestFlight went through. I think I can turn it into a PR for you this week. Send some good karma.

@CatalinVoss That would be awesome! I can't wait to see where I went wrong in my earlier attempts. Thanks!

**Comment by zaptrem** [comment]
@CatalinVoss Thank you so much for these changes! It's great that iOS support is finally almost there. However, I've run into two issues:
1. I can't seem to include bitcode once I copy the framework from the test app to my own. The error:
```

Showing All Errors Only
Bitcode bundle could not be generated because '/Users/ryantremblay/ReLearnApp/node_modules/react-native-transcription/ios/Frameworks/deepspeech_ios.framework/deepspeech_ios' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/ryantremblay/ReLearnApp/node_modules/react-native-transcription/ios/Frameworks/deepspeech_ios.framework/deepspeech_ios' for architecture arm64


```
2. I can't submit an archive of the app to App Store Connect. I get the error: 
```
ITMS-90338: Non-public API usage - The app references non-public symbols in Frameworks/deepspeech_ios.framework/deepspeech_ios: ___asan_alloca_poison, ___asan_allocas_unpoison, ___asan_handle_no_return, ___asan_init, ___asan_memset, ___asan_option_detect_stack_use_after_return, ___asan_register_image_globals, ___asan_report_load1, ___asan_report_load4, ___asan_report_load8, ___asan_report_store1, ___asan_report_store4, ___asan_report_store8, ___asan_set_shadow_f5, ___asan_shadow_memory_dynamic_address, ___asan_stack_free_5, ___asan_stack_malloc_0, ___asan_stack_malloc_1, ___asan_stack_malloc_2, ___asan_stack_malloc_3, ___asan_stack_malloc_5, ___asan_unregister_image_globals, ___asan_version_mismatch_check_apple_clang_1200. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/

```

Thanks for contributing to this issue!

**Comment by CatalinVoss** [comment]
Let's see.

On (1), hmm I'm not sure what fell apart there. I don't have too many use cases for submitting with bitcode. The main reason is for Apple to be able to rebuild your app independently of you when they make platform improvements, but the symbol mess can get complicated and as long as you update the app yourself regularly, you should be OK just disabling bitcode as I describe in the PR.

On (2), this looks like an address sanitization issue. You should be able to fix this by disabling address sanitization in the build chain?

**Comment by BryanWBear** [comment]
I get `framework not found deepspeech_ios` when building the xcworkspace folder found here: https://github.com/mozilla/DeepSpeech/tree/master/native_client/swift. 

I've tried unlinking / relinking the lib and changing the Framework search path, but no success. Does anybody have a fix for this issue?

Also another question: What's the CPU usage of this model? Would it be possible to run it on background for a long time without draining the battery, or is it more advisable to use a wakeword detection model in conjunction with it (kind of like a Siri / Alexa type setup).

**Comment by zaptrem** [comment]
@reuben I tested this on a larger file and it appears the file recognition has both a memory leak that always occurs and a crash that only occurs with larger/longer recordings. Here's a screenshot of the memory behavior and crash info with an unmodified version of the test app in the latest commit with a ~50 minute file: 
![memory usage, ](https://user-images.githubusercontent.com/1612230/102978360-626c0780-44d2-11eb-8dfc-98b303260f29.png) 
![image](https://user-images.githubusercontent.com/1612230/102978421-79aaf500-44d2-11eb-9216-ebd4e413473e.png)

Thankfully (unlike the previous Android leak I reported) it's super reliable and easy to reproduce. All you've gotta do is try to run a long audio file job.

**Comment by reuben** [comment]
We have a policy of no screenshots here, it's not accessible and hard to read. Just copy the logs over as text when reporting these things.

That stack and error looks like the input buffer passed to `feedAudioContent` is a broken pointer, how is the caller reading the audio file? Is it the unmodified `deepspeech_ios_test` project?

**Comment by zaptrem** [comment]
Away from my PC for the night/morning now, but I’ll definitely remember
that in the future.

This was completely unmodified latest commit deepspeech_ios_test (small
exception of adding the file name where the comments instructed and a
single “hello” print statement before anything runs to verify my changes to
that array were applying).

**Comment by zaptrem** [comment]
Sorry, it actually was slightly modified. I disabled Address Sanitizer and
some Zombie Detection option in order to enable memory profiling.

**Comment by zaptrem** [comment]
@reuben Enabling/disabling ASAN and Zombie Object detection makes no difference, leak and crash still appear.

**Comment by reuben** [comment]
I haven't seen conclusive evidence that there's any leak here yet, memory usage goes up over time until you hit the beam width limit. The crash is definitely weird. Can you share the audio file that triggers it?

**Comment by reuben** [comment]
Also is this with one of our pre-trained models or your own? If the latter, does the problem reproduce with our pre-trained models?

**Comment by zaptrem** [comment]
> I haven't seen conclusive evidence that there's any leak here yet, memory usage goes up over time until you hit the beam width limit. The crash is definitely weird. Can you share the audio file that triggers it?

I just emailed the address on your GitHub account. Would rather not leave a permanent link here in case it contains copyrighted material (it was shared in a bug report from a tester and I haven't listened to it).

**Comment by CatalinVoss** [comment]
There is a small race condition with the microphone capture on the stop() call in the sample code that needs to be fixed with a Semaphore that I've seen crash, but I haven't seen any other crashing behavior.

**Comment by zaptrem** [comment]
I was able to reproduce this by taking[ long speeches from Wikipedia](https://commons.wikimedia.org/wiki/Category:Audio_files_of_speeches) (e.g., https://upload.wikimedia.org/wikipedia/commons/7/7c/Reagan_at_1980_Republican_National_Convention.ogg), converting them to 16khz 1ch WAV using FFMPEG
```
ffmpeg -i file.ogg  -ar 16000 -ac 1 output.wav
```
Then running them through the example app. Weirdly enough, *none* of the speeches work at all on Android, where the transcription instantly finishes with an empty result. However, Android live transcription works fine and certain wavs (e.g., https://www2.cs.uic.edu/~i101/SoundFiles/gettysburg.wav) also work just fine. I wonder if something about the output of FFMPEG is revealing some weird bug about the way the audio is fed to the model?

Finally, when the speeches are short enough to work (~20-30 minutes?) the RAM usage on iOS is still at leaking levels. The model increased RAM usage slowly by >200MBs before finishing, whereas the file itself wasn't close to that size. (e.g., https://upload.wikimedia.org/wikipedia/commons/6/6b/D-Day_speech_%28Reagan%29.ogg). Note: that audio doesn't work at all on Android, but works on iOS.

**Comment by reuben** [comment]
The Android code does not use a robust WAV parser, it just encodes direct offsets into the file, so it struggles with inconsistent header formats, including the one generated by `ffmpeg`. You can workaround by passing `-bitexact` to `ffmpeg` when converting, hopefully that should fix it.

**Comment by zaptrem** [comment]
@reuben Thanks, that solved the Android issue! It also allowed me to compare memory usage between iOS and Android. On the D-Day speech, Android climbs to ~200mb usage almost immediately and stays there, whereas iOS climbs more slowly to a similar number (or crashes on longer speeches). It seems like you were right to assume there was no memory leak, though it's weird to me that memory use is >10x the size of the input file.

However, on longer files, the crash on iOS is still present. Memory usage on Android balloons to >500MB on the Reagan RNC speech and run for a long while (I'm virtualizing Android on a beefy PC), so I'll update if it finishes.

Additionally, (this might be related to the -bitexact issue), on iOS on the shorter speeches there are long periods of no transcription in between chunks of near-perfect transcription. The audio quality doesn't noticeably decrease in those areas when I listen. 

Finally, in the process of updating react-native-transcription to the latest DeepSpeech version I noticed that Android DeepSpeech 0.9.3 hadn't been published to Maven yet, only 0.9.2.

**Comment by zaptrem** [comment]
@reuben Another interesting bug I discovered while trying to find clues: the iOS example happily transcribed non-16khz wav files while Android example refused them. Also weird is Android and iOS give different transcriptions for the same noisy input despite using the exact same pretrained model and audio files.

**Comment by fender** [comment]
@zaptrem @CatalinVoss What's the installation process for testing on iOS? There isn't a cocoapod yet is there?

**Comment by zaptrem** [comment]
@fender You've gotta compile the framework yourself using [these ](https://github.com/mozilla/DeepSpeech/pull/3436)instructions. You can use the Swift example in the repo.

**Comment by lissyx** [comment]
> @reuben Another interesting bug I discovered while trying to find clues: the iOS example happily transcribed non-16khz wav files while Android example refused them. Also weird is Android and iOS give different transcriptions for the same noisy input despite using the exact same pretrained model and audio files.

Its not surprising and not a bug, @reuben already explained it: android example wav reading is hardcoded to reduce dependencies required by CI, so it only handles 16khz and it's not robust to broken wav files.

**Comment by zaptrem** [comment]
> I was able to reproduce this by taking[ long speeches from Wikipedia](https://commons.wikimedia.org/wiki/Category:Audio_files_of_speeches) (e.g., https://upload.wikimedia.org/wikipedia/commons/7/7c/Reagan_at_1980_Republican_National_Convention.ogg), converting them to 16khz 1ch WAV using FFMPEG
> 
> ```
> ffmpeg -i file.ogg  -ar 16000 -ac 1 output.wav
> ```
> 
> Then running them through the example app. Weirdly enough, _none_ of the speeches work at all on Android, where the transcription instantly finishes with an empty result. However, Android live transcription works fine and certain wavs (e.g., https://www2.cs.uic.edu/~i101/SoundFiles/gettysburg.wav) also work just fine. I wonder if something about the output of FFMPEG is revealing some weird bug about the way the audio is fed to the model?
> 
> Finally, when the speeches are short enough to work (~20-30 minutes?) the RAM usage on iOS is still at leaking levels. The model increased RAM usage slowly by >200MBs before finishing, whereas the file itself wasn't close to that size. (e.g., https://upload.wikimedia.org/wikipedia/commons/6/6b/D-Day_speech_%28Reagan%29.ogg). Note: that audio doesn't work at all on Android, but works on iOS.

@reuben Were you able to reproduce with this method/the file I emailed? Lmk if there's any other info I can add.

**Comment by zaptrem** [comment]
> > I was able to reproduce this by taking[ long speeches from Wikipedia](https://commons.wikimedia.org/wiki/Category:Audio_files_of_speeches) (e.g., https://upload.wikimedia.org/wikipedia/commons/7/7c/Reagan_at_1980_Republican_National_Convention.ogg), converting them to 16khz 1ch WAV using FFMPEG
> > ```
> > ffmpeg -i file.ogg  -ar 16000 -ac 1 output.wav
> > ```
> > 
> > 
> > Then running them through the example app. Weirdly enough, _none_ of the speeches work at all on Android, where the transcription instantly finishes with an empty result. However, Android live transcription works fine and certain wavs (e.g., https://www2.cs.uic.edu/~i101/SoundFiles/gettysburg.wav) also work just fine. I wonder if something about the output of FFMPEG is revealing some weird bug about the way the audio is fed to the model?
> > Finally, when the speeches are short enough to work (~20-30 minutes?) the RAM usage on iOS is still at leaking levels. The model increased RAM usage slowly by >200MBs before finishing, whereas the file itself wasn't close to that size. (e.g., https://upload.wikimedia.org/wikipedia/commons/6/6b/D-Day_speech_%28Reagan%29.ogg). Note: that audio doesn't work at all on Android, but works on iOS.
> 
> @reuben Were you able to reproduce the iOS crash with this method/the file I emailed? Lmk if there's any other info I can add.

@reuben Bumping this in case it got buried in other notifications over the holiday weekend.

**Comment by zaptrem** [comment]
@reuben Trying again in case I sent it too early in the morning last week. 

@CatalinVoss Are you able to reproduce this with your binary?

**Comment by CatalinVoss** [comment]
Can you get a backtrace from inside deepspeech by typing `bt` on the lldb console? I do see one sporadic issue that crashes within tensorflow in my application, but it's fairly rare.

**Comment by zaptrem** [comment]
@CatalinVoss Here you go. I had to use a text file since I exceeded pastebin's size limit: [deepSpeechCrashST.txt](https://github.com/mozilla/DeepSpeech/files/5806797/deepSpeechCrashST.txt)

**Comment by CatalinVoss** [comment]
Ah. The relevant part is at the bottom

```
deepspeech_ios`PathTrie::iterate_to_vec(std::__1::vector<PathTrie*, std::__1::allocator<PathTrie*> >&) + 64
    frame #4742: 0x000000010fee1114 deepspeech_ios`DecoderState::next(double const*, int, int) + 404
    frame #4743: 0x000000010fd4b470 deepspeech_ios`StreamingState::processBatch(std::__1::vector<float, std::__1::allocator<float> > const&, unsigned int) + 296
    frame #4744: 0x000000010fd4b300 deepspeech_ios`StreamingState::pushMfccBuffer(std::__1::vector<float, std::__1::allocator<float> > const&) + 236
    frame #4745: 0x000000010fd4ae88 deepspeech_ios`StreamingState::feedAudioContent(short const*, unsigned int) + 396
  * frame #4746: 0x0000000105f1c808 ReLearn`closure #1 in Transcription.render(samples=Swift.UnsafeRawBufferPointer @ 0x000000016b295c30, stream=(streamCtx = 0x000000011031e420)) at Transcription.swift:85:24
    frame #4747: 0x0000000105f1c830 ReLearn`thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #4748: 0x0000000105f1e3d0 ReLearn`partial apply for thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #4749: 0x00000001f031ca8c libswiftFoundation.dylib`Foundation.Data.withUnsafeBytes<A>((Swift.UnsafeRawBufferPointer) throws -> A) throws -> A + 504
    frame #4750: 0x0000000105f1bf68 ReLearn`Transcription.render(audioContext=0x0000000280435b30, stream=(streamCtx = 0x000000011031e420), self=0x00000002805e2130) at Transcription.swift:83:26
    frame #4751: 0x0000000105f1ceb8 ReLearn`closure #1 in Transcription.recognizeFile(audioContext=0x0000000280435b30, self=0x00000002805e2130, stream=(streamCtx = 0x000000011031e420), audioPath=Swift.String @ 0x000000016b2964a8, start=2021-01-13 01:27:08 EST) at Transcription.swift:107:18
    frame #4752: 0x0000000105f1a23c ReLearn`closure #1 in static AudioContext.load(asset=0x0000000280c2a500, assetTrack=0x00000002808f4420, audioURL=<unavailable; try printing with "vo" or "po">, completionHandler=0x0000000105f1e48c ReLearn`partial apply forwarder for closure #1 (Swift.Optional<react_native_transcription.AudioContext>) -> () in react_native_transcription.Transcription.(recognizeFile in _79B1BC2F893AB0135086535C16DBA135)(audioPath: Swift.String) -> () at <compiler-generated>) at AudioContext.swift:58:17
    frame #4753: 0x0000000105ef6490 ReLearn`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0
    frame #4754: 0x00000001101c9d10 libdispatch.dylib`_dispatch_call_block_and_release + 32
    frame #4755: 0x00000001101cb18c libdispatch.dylib`_dispatch_client_callout + 20
    frame #4756: 0x00000001101d2968 libdispatch.dylib`_dispatch_lane_serial_drain + 724
    frame #4757: 0x00000001101d3580 libdispatch.dylib`_dispatch_lane_invoke + 440
    frame #4758: 0x00000001101df0f0 libdispatch.dylib`_dispatch_workloop_worker_thread + 1344
    frame #4759: 0x00000001b96f3714 libsystem_pthread.dylib`_pthread_wqthread + 276
(lldb) 
```

This looks like a crash in the beam search decoder. I'm not actually using the DeepSpeech decoder -- have my own -- so unfortunately I can't help :( it appears to be unrelated to the bug I saw

**Comment by zaptrem** [comment]
@CatalinVoss Thanks for pointing that out! I'm curious; what about your use case required building an alternate decoder? Is it open source? 

@reuben What do you make of this error? Or is there a different person who worked on the beam decoder I should ping? Is this something I should move to its own, non-platform-specific bug report?

**Comment by CatalinVoss** [comment]
> @CatalinVoss Thanks for pointing that out! I'm curious; what about your use case required building an alternate decoder? Is it open source? 

Working on child literacy. Closed source I'm afraid

**Comment by zaptrem** [comment]
@CatalinVoss Cool stuff! I can see why it's necessary in that case, detecting specific mispronunciations is more than hotword-boosting can do. Coincidentally, I stumbled across this [Google feature that might interest you](https://www.theverge.com/2019/11/14/20964401/google-search-pronunciation-guide-feedback-machine-learning-ai) about 10 minutes ago in a discussion with someone.

I ran the test again on an iPad Pro late 2018 (before it was an iPhone 11 Pro Max) and got a slightly different result. Does this still indicate Decoder issues?
```
frame #4739: 0x000000010bb6bcf8 deepspeech_ios`PathTrie::iterate_to_vec(std::__1::vector<PathTrie*, std::__1::allocator<PathTrie*> >&) + 64
    frame #4740: 0x000000010bb6bcf8 deepspeech_ios`PathTrie::iterate_to_vec(std::__1::vector<PathTrie*, std::__1::allocator<PathTrie*> >&) + 64
    frame #4741: 0x000000010ba99114 deepspeech_ios`DecoderState::next(double const*, int, int) + 404
    frame #4742: 0x000000010b903470 deepspeech_ios`StreamingState::processBatch(std::__1::vector<float, std::__1::allocator<float> > const&, unsigned int) + 296
    frame #4743: 0x000000010b903300 deepspeech_ios`StreamingState::pushMfccBuffer(std::__1::vector<float, std::__1::allocator<float> > const&) + 236
    frame #4744: 0x000000010b902e88 deepspeech_ios`StreamingState::feedAudioContent(short const*, unsigned int) + 396
  * frame #4745: 0x0000000101a80808 ReLearn`closure #1 in Transcription.render(samples=Swift.UnsafeRawBufferPointer @ 0x000000017032dc20, stream=(streamCtx = 0x000000010bfd2670)) at Transcription.swift:85:24
    frame #4746: 0x0000000101a80830 ReLearn`thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #4747: 0x0000000101a823d0 ReLearn`partial apply for thunk for @callee_guaranteed (@unowned UnsafeRawBufferPointer) -> (@error @owned Error) at <compiler-generated>:0
    frame #4748: 0x00000001942f5174 libswiftFoundation.dylib`Foundation.Data.withUnsafeBytes<A>((Swift.UnsafeRawBufferPointer) throws -> A) throws -> A + 392
    frame #4749: 0x0000000101a7ff68 ReLearn`Transcription.render(audioContext=0x0000000282e94930, stream=(streamCtx = 0x000000010bfd2670), self=0x0000000282f28cc0) at Transcription.swift:83:26
    frame #4750: 0x0000000101a80eb8 ReLearn`closure #1 in Transcription.recognizeFile(audioContext=0x0000000282e94930, self=0x0000000282f28cc0, stream=(streamCtx = 0x000000010bfd2670), audioPath=Swift.String @ 0x000000017032e458, start=2021-01-13 02:00:54 EST) at Transcription.swift:107:18
    frame #4751: 0x0000000101a7e23c ReLearn`closure #1 in static AudioContext.load(asset=0x000000028278c8a0, assetTrack=0x00000002822ea6d0, audioURL=<unavailable; try printing with "vo" or "po">, completionHandler=0x0000000101a8248c ReLearn`partial apply forwarder for closure #1 (Swift.Optional<react_native_transcription.AudioContext>) -> () in react_native_transcription.Transcription.(recognizeFile in _79B1BC2F893AB0135086535C16DBA135)(audioPath: Swift.String) -> () at <compiler-generated>) at AudioContext.swift:58:17
    frame #4752: 0x0000000101a5a490 ReLearn`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0
    frame #4753: 0x000000010bd83bcc libdispatch.dylib`_dispatch_call_block_and_release + 32
    frame #4754: 0x000000010bd856c0 libdispatch.dylib`_dispatch_client_callout + 20
    frame #4755: 0x000000010bd8d354 libdispatch.dylib`_dispatch_lane_serial_drain + 736
    frame #4756: 0x000000010bd8e0c0 libdispatch.dylib`_dispatch_lane_invoke + 448
    frame #4757: 0x000000010bd9a644 libdispatch.dylib`_dispatch_workloop_worker_thread + 1520
    frame #4758: 0x00000001db297804 libsystem_pthread.dylib`_pthread_wqthread + 276
```

**Comment by CatalinVoss** [comment]
Oh cool I had not seen that!

Yes this still looks like decoder. :/

**Comment by zaptrem** [comment]
> > Yes, I'm already doing that because the Metadata and Token arrays were (erroneously?) marked as `private(set)` instead of `public private(set)`.
> 
> PRs are welcome.

@reuben It's a bit late, but I just proposed those quick changes here: https://github.com/mozilla/DeepSpeech/pull/3510

I also think there's some more debug info I can provide about the decoder issue, I'll test it tomorrow.

**Comment by zaptrem** [comment]
@reuben  Okay, poking at Xcode's debugging options/info is like trying to reverse engineer an alien spaceship at my current knowledge level, so I haven't learned much from a few hours trying that. 

*However,* I did get some interesting results from messing around with the scorer/models. Disabling the external scorer allowed the transcription process to run for much longer before eventually crashing (similar to the length it took my Android build to complete the same transcription task) with the same error. Additionally, using the Chinese model and scorer lasted the same extended amount of time as disabling the external scorer. However, the "reading samples" message varied exactly once near the beginning of the job like so:

```
...
reading 8192 samples
reading 7425 samples
reading 8192 samples
...
```

**Comment by zaptrem** [comment]
@CatalinVoss Any idea why it would still be failing when the external scorer is disabled?

**Comment by zaptrem** [comment]
@reuben Does any of this help? It's been over a month since your last response. Is there potentially an FFMPEG setting that is causing this, since it doesn't happen with non-converted WAV files?  Is there any more experimentation/testing I can do to help this along?

**Comment by zaptrem** [comment]
@reuben I tried converting from MKV files encoded with vorbis instead of the .ogg I was using before. Didn't work. I tried converting with Audacity manually. Didn't work. 

I wonder if it has nothing to do with the conversion and it's actually just the length of the recording. Do you know of any long conversations recorded in PCM 16 natively we can test with? 

I'm lead to this conclusion not just from my testing, but because @CatalinVoss  suggested it's a decoder issue (implying to me the audio file had already been successfully read/inference had completed).

**Comment by zaptrem** [comment]
@CatalinVoss (and @reuben if you're still here) I continued testing and I think the issue might(????) have something to do a bug in render() caused by however FFMPEG converts 48000mhz to 16000mhz.

I rewrote the render function in iOS to be simpler and (I would think) more robust due to the utilization of higher-level Apple APIs that are file-format independent. The crash disappeared (hooray!) but I'm now getting jibberish short transcriptions (whereas the same files on Android give me acceptable results). Any idea where I went wrong here?: 

```swift
private func newRender(url: URL, stream: DeepSpeechStream) {
        let file = try! AVAudioFile(forReading: url)
        guard let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: file.fileFormat.sampleRate, channels: 1, interleaved: false) else { fatalError("Couldn't read the audio file") }
        print("reading file")
        var done = false;
        while(!done){
            let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: 8096)!
            do {
                try file.read(into: buf)
            } catch {
                print("end of file");
                done = true
                return;
            }
            print("read 8096 frames")
            let floatBufPointer = UnsafeBufferPointer(start: buf.int16ChannelData?[0], count: Int(buf.frameLength))
            stream.feedAudioContent(buffer: floatBufPointer);
        }
    }

```

**Comment by CatalinVoss** [comment]
Sorry for the delay.

I never used the audio file recognition path that goes through `render()` so an error there could explain why you're facing issues and I'm not. I'm just capturing mic output.

As a hackky workaround, are you just calling `finishStream()` on the stream once at the very end? Is there any way you can split your monsterous audio file into chunks, passing a few chunks at at a time and "finalize" the stream a few times in between?

Another way to isolate the issue may be to see if you encounter the issue with long-running mic detection (which, again, I don't see), but I'm not doing hours and hours of it either.

As for your `newRender()`, hard to debug without playing with it. This kinda stuff is tricky to get right, but you only have to do it once. I am not sure why you're calling this guy a `floatBufPointer`, since the stream takes 16-bit int samples. With what you'e doing now, it certainly looks like you need to be 100% sure that your input is PCM single channel at 16 kHz.

**Comment by zaptrem** [comment]
@CatalinVoss Thanks for getting back!! I could split it into chunks, but it would likely mess with transcriptions mid-sentence and paragraph, as well as reset the time part of tokens (I'm using intermediateDecodeWithMetadata as finishStreamWithMetadata also has an unknown crash... one problem at a time). 

It's called floatBufPointer because originally I was passing floats instead of 16bit ints and I forgot to change it. As you can see here:
```swift
 guard let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: file.fileFormat.sampleRate, channels: 1, interleaved: false) else { fatalError("Couldn't read the audio file") }

...

let floatBufPointer = UnsafeBufferPointer(start: buf.int16ChannelData?[0], count: Int(buf.frameLength))
```
I am passing the int16ChannelData. TypeOf tells me this is a pointer to an int16 array. I can also wrap it in a swift Array() and typeOf is int16[]. Weirdly enough, when I pass the Swift Array() I get a different (but still jibberish) result. 

Would it be helpful if I opened a pull request with these changes so they're easy for others to look at?

**Comment by zaptrem** [comment]
@reuben Any idea what I’m doing wrong in the proposed Swift code above?

**Comment by zaptrem** [comment]
@reuben Can you respond with an estimate on when you can look at this? It's been over a month and a half. I've given it my best shot but have run out of ideas for now.

**Comment by zaptrem** [comment]
@reuben @lissyx I gave up on fixing this and just took the accuracy hit from splitting the audio file into 5-minute segments. I've published the app using this, [ReLearn](https://relearn.fyi), to the Android/iOS App Stores. It uses DeepSpeech to transcribe long video recordings of lectures for free on-device in the background (it also transcribes audio in-person recordings on Android, but uses Apple's solution for that on iOS for now).

**Comment by martin642** [comment]
Fantastic job

**Comment by lustig-bakkt** [comment]
> @reuben @lissyx I gave up on fixing this and just took the accuracy hit from splitting the audio file into 5-minute segments. I've published the app using this, [ReLearn](https://relearn.fyi), to the Android/iOS App Stores. It uses DeepSpeech to transcribe long video recordings of lectures for free on-device in the background (it also transcribes audio in-person recordings on Android, but uses Apple's solution for that on iOS for now).

@zaptrem I've been wanting to build something very similar to this but just ran up against Apple's one minute live transcription limit with SFSpeechRecognizer. How are you getting around this in ReLearn? Any chance you'd like to collaborate on some live speech / mind mapping type applications?
